src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -280,29 +295,119 @@ def process_input(conf):
             break
         upa, obj_info, genome_upa = task
 
-        # cfn points to the assembly file outside of the container
-        # get_assembly_as_fasta writes the file to /kb/module/workdir/tmp/<filename> inside the container.
-        # workdir is shared between the container and the external file system
-        # Any file path get_assembly_as_fasta returns will be relative to inside the container, and so is not useful for this script
+        upa_dir = os.path.join(conf.pth, upa)
+        metafile = os.path.join(upa_dir, f"{upa}.meta")
+        if not os.path.isdir(upa_dir) or not loader_helper.is_upa_info_complete(upa_dir):
+
+            # remove legacy upa_dir to avoid FileExistsError in hard link
+            if os.path.isdir(upa_dir):
+                shutil.rmtree(upa_dir)
+
+            # cfn points to the assembly file outside of the container
+            # get_assembly_as_fasta writes the file to /kb/module/workdir/tmp/<filename> inside the container.
+            # workdir is shared between the container and the external file system
+            # Any file path get_assembly_as_fasta returns will be relative to inside the container, and so is not useful for this script
+            cfn = os.path.join(conf.job_dir, "workdir/tmp", upa)
+            # upa file is downloaded to cfn
+            conf.asu.get_assembly_as_fasta({"ref": upa.replace("_", "/"), "filename": upa})
+
+            # each upa in output_dir as a separate directory
+            os.makedirs(upa_dir, exist_ok=True)
+
+            dst = os.path.join(upa_dir, f"{upa}.fa")
+            # Hard link .fa file from job_dir to output_dir in WS
+            os.link(cfn, dst)
+
+            # save meta file with relevant object_info
+            with open(metafile, "w", encoding="utf8") as json_file:
+                json.dump(_process_object_info(obj_info, genome_upa), json_file, indent=2)
+
+            print("Completed %s" % (upa))
+        else:
+            print(f"Skip downloading {upa} as it already exists")
+
+        if conf.retrieve_sample:
+            _download_sample_data(conf, upa, metafile)
+
+
+def _download_sample_data(conf: Conf, upa: str, metafile: str) -> None:
```

Comment:
This method is ~50 lines now, can it be split up?

Reply:
👍

Reply:
Hmm, it's still 40 lines, but it's reasonably readable... up to you if you want to try and further split things up